spec/queries/get_activity_log_events_spec.rb

```diff
@@ -0,0 +1,122 @@
+      array = 2.times.map { create_application_choice_for_course(course_provider_a) }
+      expect { GetActivityLogEvents.new(application_choices: array).call }.to raise_error(NoMethodError)
+    end
+
+    it 'returns an empty array if no audits are found' do
+      expect(service.call).to eq([])
+    end
+
+    it 'returns objects responding to the required attributes' do
+      choice = create_application_choice_for_course course_provider_a
+      create_audit_for_application_choice choice
+
+      result = service.call
+
+      expect(result.count).to eq(1)
+
+      %i[created_at auditable].each do |attr|
+        expect(result.first).to respond_to(attr)
+      end
+    end
+
+    it 'supports an since: argument for limiting number of events' do
+      choice = create_application_choice_for_course course_provider_a
+      create_audit_for_application_choice choice
+
+      expected = create(
+        :application_choice_audit,
+        user: provider_user,
+        application_choice: choice,
+        created_at: @now + 1.day,
+      )
+
+      result = service.call(since: @now + 6.hours)
+
+      expect(result.first).to eq(expected)
+    end
+  end
+
+  context 'includes all and only relevant audits' do
+    it 'filters on action == update' do
+      choice = create_application_choice_for_course course_provider_a
+
+      result = service.call
+
+      expect(choice.audits.count).to eq(1)
+      expect(choice.audits.first.action).to eq('create')
+      expect(result.count).to eq(0)
+    end
+  end
+
+  context 'sorts events in reverse chronological order' do
+    it 'within an application' do
+      choice = create_application_choice_for_course course_provider_a
+      audits = 3.times.map { create_audit_for_application_choice choice }
+
+      result = service.call
+
+      expect(result.map(&:id)).to eq(audits.reverse.map(&:id))
+    end
+
+    it 'across applications and providers' do
+      choice_a = create_application_choice_for_course course_provider_a
+      choice_b = create_application_choice_for_course course_provider_b
+
+      audits = 5.times.map do
+        create_audit_for_application_choice [choice_a, choice_b].sample
+      end
+
+      result = service.call
+
+      expect(result.map(&:id)).to eq(audits.reverse.map(&:id))
+    end
+  end
+
+  context 'completes in a reasonable timeframe' do
+    it '<50ms for 1000 application choices' do
+      skip 'This spec takes a long time and should be run manually'
+
+      1000.times do
+        %i[course_provider_a course_provider_b course_unrelated ratified_course_provider_b ratified_course_unrelated].each do |course|
+          choice = create_application_choice_for_course send(course)
+          20.times.map { create_audit_for_application_choice choice }
+        end
+      end
+
+      elapsed_time = Benchmark.measure { service.call }.real
+      puts "GetProviderActivityLogEvents #call completed in #{elapsed_time}s"
+
+      expect(elapsed_time).to be < 0.05
```

Comment:
I'm not sure of the benefit of running this as part of the test suite by default - especially taking into account how slow our test suite already issue.

Maybe it's worth adding a new performance check as part of the repository's GitHub actions, and tagging and excluding this from the default run, so in case there are any concerns the issue can be both noticed and also addressed as a separate issue.

Reply:
True but we don't have support for automated tests against deployed environments yet (well, apart for smoke tests, but these won't help us in this case). I'd be reluctant to run this or similar code on one of our permanent deployed environments, because it will generate lots of junk. The advantage of running this within a test context is the database cleanup.

Reply:
Sorry Despo, missed your comment. This scenario is meant to be run manually and is actually skipped for normal rspec runs. Perhaps I should just remove it.